### ISSUE TITLE
t.web on Python 3 doesn't handle CGI multipart very well

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8052]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#8052|
|Type|defect|
|Created|2015-10-19 15:51:11Z|
|Branch|https://github.com/twisted/twisted/tree/multipart-py3-8052|

From https://github.com/crossbario/crossbar/issues/481:

```
2015-10-15T10:09:15-0400 [Router      21661] Unhandled Error
Traceback (most recent call last):
  File "/usr/local/lib/python3.4/dist-packages/twisted/python/log.py", line 101, in callWithLogger
    return callWithContext({"system": lp}, func, *args, **kw)
  File "/usr/local/lib/python3.4/dist-packages/twisted/python/log.py", line 84, in callWithContext
    return context.call({ILogContext: newCtx}, func, *args, **kw)
  File "/usr/local/lib/python3.4/dist-packages/twisted/python/context.py", line 118, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/usr/local/lib/python3.4/dist-packages/twisted/python/context.py", line 81, in callWithContext
    return func(*args,**kw)
--- <exception caught here> ---
  File "/usr/local/lib/python3.4/dist-packages/twisted/internet/posixbase.py", line 597, in _doReadOrWrite
    why = selectable.doRead()
  File "/usr/local/lib/python3.4/dist-packages/twisted/internet/tcp.py", line 209, in doRead
    return self._dataReceived(data)
  File "/usr/local/lib/python3.4/dist-packages/twisted/internet/tcp.py", line 215, in _dataReceived
    rval = self.protocol.dataReceived(data)
  File "/usr/local/lib/python3.4/dist-packages/twisted/protocols/basic.py", line 578, in dataReceived
    why = self.rawDataReceived(data)
  File "/usr/local/lib/python3.4/dist-packages/twisted/web/http.py", line 1763, in rawDataReceived
    self._transferDecoder.dataReceived(data)
  File "/usr/local/lib/python3.4/dist-packages/twisted/web/http.py", line 1436, in dataReceived
    finishCallback(data[contentLength:])
  File "/usr/local/lib/python3.4/dist-packages/twisted/web/http.py", line 1694, in _finishRequestBody
    self.allContentReceived()
  File "/usr/local/lib/python3.4/dist-packages/twisted/web/http.py", line 1757, in allContentReceived
    req.requestReceived(command, path, version)
  File "/usr/local/lib/python3.4/dist-packages/twisted/web/http.py", line 810, in requestReceived
    args.update(cgi.parse_multipart(self.content, pdict))
  File "/usr/lib/python3.4/cgi.py", line 235, in parse_multipart
    nextpart = b"--" + boundary
builtins.TypeError: can't concat bytes to str
```

<details><summary>Searchable metadata</summary>

```
trac-id__8052 8052
type__defect defect
reporter__hawkowl hawkowl
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_multipart_py3_8052 branches/multipart-py3-8052
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__web web
keywords__None None
time__1445269871197845 1445269871197845
changetime__1445514210897205 1445514210897205
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_1"></a>|@hawkowl commented|
|-|-|

(In [45873]) Branching to multipart-py3-8052.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl commented|
|-|-|

Builds are looking good. Please review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Changes looks good.

Why is the pdict encoding required in _parseHeaders ?

test_contentDisposition is only about content-disposition handling or is also about using the content-disposition together with multipart ?

Please check my comments and merge :)

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_4"></a>|@hawkowl commented|
|-|-|

Thanks Adi, I've made those changes, will merge.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl set status to `closed`|
|-|-|

(In [45893]) Merge multipart-py3-8052: Fix multipart handling on Python 3

Author: hawkowl
Reviewer: adiroiban
Fixes: [#8052](https://github.com/twisted/twisted/issues/8052)
